Theme Info Sheet: Add beta badge to FSE block themes

jeyip opened the pull request:
Dependent on code refactored in https://github.com/Automattic/wp-calypso/pull/56123

### Changes proposed in this Pull Request
Add a Beta label to the theme info sheet title banner for block-based themes. Also makes the "Activate this design" button a primary button as requested by @onuro.

### Screenshots
#### Before
<img width="1278" alt="Screen Shot 2021-09-13 at 3 09 26 PM" src="https://user-images.githubusercontent.com/5414230/133163122-85831bae-8c07-4876-a2c3-8e21854307d5.png">

#### After
<img width="1267" alt="Screen Shot 2021-09-15 at 10 51 50 AM" src="https://user-images.githubusercontent.com/5414230/133484499-a2d8715c-b739-4b46-b282-e87bc13b34ba.png">

### Testing instructions

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

- Open Appearance -> Themes
- Confirm current themes that support block-templates have a Beta badge

<!--
Link a related issue to this PR. If the PR does not immediately resolve the issue,
for example, it requires a separate deployment to production, avoid
using the "fixes" keyword and instead attach the [Status] Fix Inbound label to
the linked issue.
-->

Related to https://github.com/Automattic/wp-calypso/issues/54508

jeyip committed:
Add beta badge to theme info sheet title

jeyip reviewed (commented):
**Note:**

I can place the beta badge anywhere on the page, but if we want it next to the theme info sheet title, we may want to choose a different color @onuro. 

It's not the most a11y friendly on a dark background.

<img width="810" alt="Screen Shot 2021-09-13 at 3 16 30 PM" src="https://user-images.githubusercontent.com/5414230/133163648-ab5d33d7-511e-4f8a-a621-c40854d236e4.png">

jeyip commented:
Apologies for the ping everyone! Bad rebase. I untagged all unnecessary reviewers.

Addison-Stavlo reviewed (commented):
Tested this out and it works as expected!  

The same concern I noted (https://github.com/Automattic/wp-calypso/pull/56123#discussion_r708225073) regarding showing the beta badge on universal themes exists here, but other than that I think this is good.

We can choose to update that here in the `isFullSiteEditingTheme` function, or create a follow-up ticket and be sure to update it before Universal themes are released to non-beta users.

jeyip committed:
Add white background to beta badge

jeyip committed:
Make Activate design button a primary button

Copons commented on client/my-sites/theme/main.jsx:
```diff
@@ -613,11 +622,13 @@ class ThemeSheet extends React.Component {
 				{ this.isLoaded() ? label : placeholder }
-				{ this.props.isWpcomTheme && this.renderPrice() }
+				<Badge type="info" className="theme__sheet-badge-beta">
+					{ this.props.isWpcomTheme && this.renderPrice() }
```
I've noticed a fun issue with this.

If the theme we are looking at is the active one for the current site, `renderPrice()` renders an empty string.

Previously, this was ok, since we were rendering it "unwrapped".
Now, though, it's inside a `Badge`.
This is what it looks like when `renderPrice()` is empty:

<img width="190" alt="Screenshot 2021-09-28 at 16 29 46" src="https://user-images.githubusercontent.com/2070010/135118554-23097b21-962d-458f-a0f0-0fca19bd427b.png">

jeyip replied:
Good catch! I'll take a look in a bit.

jeyip replied:
Made updates with in f3f6e02

Copons commented on client/utils.js:
```diff
@@ -7,3 +7,8 @@ export function pathToRegExp( path ) {
+export function isFullSiteEditingTheme( theme ) {
+	const features = theme?.taxonomies?.theme_feature;
+	return features && features.some( ( feature ) => feature.slug === 'block-templates' );
+}
```
This file, am I right? It's older than me!

I kinda feel like this is not the most appropriate place to add a utility function — and I'm sort of backed by the fact that nobody touched it in years! 😄 

What about the `client/lib` folder?
It's full of single-function folders, but I would be ok if we created a `full-site-editing` folder that could maybe contain several FSE-related utilities in the future.

Alternatively, and arguably even better, it would be cool to have this function closer to where it's used.
I presume that you moved it here because it's used by two different routes (`themes` and `theme`), but it happens quite often for a route to import stuff from another.
For example, I can see lots of components from `themes` used in `theme`.
What about moving it into a new `client/my-sites/themes/is-full-site-editing-theme.js` file?

jeyip replied:
> I presume that you moved it here because it's used by two different routes (themes and theme)

Yep that's exactly why.

> What about moving it into a new client/my-sites/themes/is-full-site-editing-theme.js file?

> it happens quite often for a route to import stuff from another.

Happy to do this. I find it to be more intuitive to have shared functions live in the closest possible _ancestor_ in the directory structure, but I recognize that it's an opinion, and I'm totally fine following existing conventions. I'll make the changes.

Copons replied:
Yeah I know what you mean.
I'm rusty with Calypso conventions at this point, so I'm just guessing that, given there are no shared utilities in the ancestor (`client/my-sites`), and given the function seems a bit too specific to be included in the main utilities folder (`client/lib`),  it _should_ be fine to just put it into one between `themes` and `theme` and call it a day.

But most importantly: I'd avoid `client/utils.js`. 😄 
I don't have any historical knowledge about it, but I'd argue that such a generic file, if left unchecked, could have easily spun out of control, getting chock full of random one-off functions.

jeyip replied:
Ah and I just realized, like you pointed out, that `client/my-sites` would've been the best place to put the shared function given my preferences 😅

Either way, I'll go ahead and move things to a new `client/my-sites/themes/is-full-site-editing-theme.js`

jeyip replied:
Done in 8807b10

jeyip committed:
Update full site editing theme validation function import statement

jeyip committed:
Hide button badge when price string is empty

When the price string is empty, such as when a user views information about the currently active theme, displaying a price badge becomes unnecessary.

Copons commented on client/my-sites/themes/current-theme/index.jsx:
```diff
@@ -126,5 +147,6 @@ export default connect( ( state, { siteId } ) => {
 	return {
 		currentThemeId,
 		currentTheme: getCanonicalTheme( state, siteId, currentThemeId ),
+		isSiteUsingCoreSiteEditor: isSiteUsingCoreSiteEditorSelector( state, siteId ),
```
We could straight up return `showBetaBadge` here instead of computing it in the rendered.
E.g.

```js
const currentTheme = getCanonicalTheme( state, siteId, currentThemeId );
return {
  currentTheme,
  showBetaBadge: isFullSiteEditingTheme( currentTheme ) && isSiteUsingCoreSiteEditor( state, siteId )
}
```

It would also improve the import naming (`isSiteUsingCoreSiteEditorSelector` makes it slightly harder to search), given we wouldn't destructure the prop anymore.

Copons reviewed (approved):
LGTM! ✨ 

The comment is a code style nitpick, feel free to ignore it!

vindl commented on client/my-sites/theme/main.jsx:
```diff
@@ -613,11 +623,15 @@ class ThemeSheet extends React.Component {
 				className="theme__sheet-primary-button"
 				href={ getUrl ? getUrl( this.props.id ) : null }
 				onClick={ this.onButtonClick }
-				primary={ isActive }
```
Why are we changing this to be `primary` unconditionally?

jeyip replied:
If it's conditional, then, when looking at any theme that isn't active, the button will appear like so.

<img width="1014" alt="Screen Shot 2021-10-06 at 3 44 09 PM" src="https://user-images.githubusercontent.com/5414230/136293882-d8b42980-511d-4149-93be-2ce166457a8c.png">

Making the "Activate this design" button a primary button was a request from @onuro p1631698375353200-slack-CHN6J22MP because users seemed to have a hard time locating it otherwise.
